File: DSA/Ansh-Gulati/Day-17/Day-17_Question-2.cpp
#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

TreeNode* insertIntoBST(TreeNode* root, int val) {
    if (!root) {
        // If the tree is empty, create a new node and return it as the root
        return new TreeNode(val);
    }

    if (val < root->val) {
        // If the value to be inserted is less than the current node's value, insert it into the left subtree
        root->left = insertIntoBST(root->left, val);
    } else {
        // If the value to be inserted is greater than or equal to the current node's value, insert it into the right subtree
        root->right = insertIntoBST(root->right, val);
    }

    // Return the root node after insertion
    return root;
}

void inorderTraversal(TreeNode* root) {
    if (!root) return;
    inorderTraversal(root->left);
    cout << root->val << " ";
    inorderTraversal(root->right);
}

int main() {
    TreeNode* root = new TreeNode(4);
    root->left = new TreeNode(2);
    root->right = new TreeNode(7);
    root->left->left = new TreeNode(1);
    root->left->right = new TreeNode(3);

    int val = 5;
    cout << "BST before insertion: ";
    inorderTraversal(root);
    cout << endl;

    root = insertIntoBST(root, val);

    cout << "BST after insertion: ";
    inorderTraversal(root);
    cout << endl;

    return 0;
}

File: DSA/Vikash_kumar/Day_17_Question_2.py
Day17-Question2-
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def insertIntoBST(root, val):
    if root is None:
        return TreeNode(val)
      
    current = root
    while True:
        if val < current.val:
            if current.left is None:
                current.left = TreeNode(val)
                break
            else:
                current = current.left
        else: 
            if current.right is None:
                current.right = TreeNode(val)
                break
            else:
                current = current.right

    return root

root = TreeNode(4)
root.left = TreeNode(2)
root.right = TreeNode(7)
root.left.left = TreeNode(1)
root.left.right = TreeNode(3)

new_root = insertIntoBST(root, 5)

def inorder_traversal(node):
    if node:
        inorder_traversal(node.left)
        print(node.val, end=' ')
        inorder_traversal(node.right)

inorder_traversal(new_root)
